metadata-ingestion/src/datahub/ingestion/source/hex/mapper.py

```diff
@@ -0,0 +1,332 @@
+        self._set_ownership_from_email = set_ownership_from_email
+
+    def map_workspace(self) -> Iterable[MetadataWorkUnit]:
+        container_properties = ContainerPropertiesClass(
+            name=self._workspace_name,
+        )
+        aspects: List[_Aspect] = [container_properties]
+        yield from self._yield_mcps(
+            entity_urn=self._workspace_urn,
+            aspects=aspects,
+        )
+
+    def map_project(self, project: Project) -> Iterable[MetadataWorkUnit]:
+        dashboard_urn = self._get_dashboard_urn(name=project.id)
+
+        dashboard_info = DashboardInfoClass(
+            title=project.title,
+            description=project.description or "",
+            lastModified=self._change_audit_stamps(
+                created_at=project.created_at, last_edited_at=project.last_edited_at
+            ),
+            externalUrl=f"{self._base_url}/{self._workspace_name}/hex/{project.id}",
+            customProperties=dict(id=project.id),
+        )
+
+        subtypes = SubTypesClass(
+            typeNames=[BIAssetSubTypes.HEX_PROJECT],
+        )
+
+        platform_instance = self._platform_instance_aspect()
+
+        container = ContainerClass(
+            container=self._workspace_urn.urn(),
+        )
+
+        aspects = [dashboard_info, subtypes, platform_instance, container]
+
+        tags = self._global_tags(
+            status=project.status,
+            categories=project.categories,
+            collections=project.collections,
+        )
+        if tags:
+            aspects.append(tags)
+
+        ownership = self._ownership(creator=project.creator, owner=project.owner)
+        if ownership:
+            aspects.append(ownership)
+
+        dashboard_usage_statistics = self._dashboard_usage_statistics(
+            analytics=project.analytics
+        )
+        if dashboard_usage_statistics:
+            aspects.append(dashboard_usage_statistics)
+
+        yield from self._yield_mcps(entity_urn=dashboard_urn, aspects=aspects)
+
+    def map_component(self, component: Component) -> Iterable[MetadataWorkUnit]:
+        dashboard_urn = self._get_dashboard_urn(name=component.id)
+
+        dashboard_info = DashboardInfoClass(
+            title=component.title,
+            description=component.description or "",
+            lastModified=self._change_audit_stamps(
+                created_at=component.created_at, last_edited_at=component.last_edited_at
+            ),
+            externalUrl=f"{self._base_url}/{self._workspace_name}/hex/{component.id}",
+            customProperties=dict(id=component.id),
+        )
+
+        subtypes = SubTypesClass(
+            typeNames=[BIAssetSubTypes.HEX_COMPONENT],
+        )
+
+        platform_instance = self._platform_instance_aspect()
+
+        container = ContainerClass(
+            container=self._workspace_urn.urn(),
+        )
+
+        aspects = [dashboard_info, subtypes, platform_instance, container]
```

Comment:
if feels strange to create an aspects array halfway through the method

note that `MetadataChangeProposalWrapper.construct_many` will transparently skip `None`s. so you can do `.construct_many(urn, aspects=[..., tags, ownership, ...]` even if tags is of type `Optional[GlobalTagsClass]`

Reply:
I was trying to avoid adding `None` objects to the aspects array, if `construct_many` skips them, yes, I can simplify that

addressed in commit 403af0a6d4fadb1b7a8f099e92deb040bbb99d0a